common/src/main/java/io/pravega/common/util/AbstractBufferView.java

```diff
@@ -0,0 +1,200 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.hash.HashHelper;
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+import lombok.Getter;
+import lombok.SneakyThrows;
+
+/**
+ * Base implementation of {@link BufferView}, providing common functionality.
+ * Derived classes may override these methods with more efficient implementations tailored to their data types.
+ */
+public abstract class AbstractBufferView implements BufferView {
+    static final BufferView EMPTY = new EmptyBufferView();
+    private static final HashHelper HASH = HashHelper.seededWith(AbstractBufferView.class.getName());
+
+    @Override
+    public int hashCode() {
```

Comment:
The goal of this is that any `BufferView` implementation would have the same hashcode and have `equals(...)` return true if their payload is the same. So I had to come up with a common hash code algorithm that works for all.

If there are more optimized solutions that would keep this working, I'd be up to fix them. However I compared this with `ByteBuffer`, `ByteBuf` and the old `HashedArray` implementations and they all so something equally "slow". That's because they need to account for the fact that the buffer payload may change between invocations, so the hash code must change accordingly.

FYI we could do this later as these hashcodes are just used in memory and they're not persisted; there's no backwards compatibility we need to worry about this.

Reply:
Yes this is only for in-memory equality and hashes. So any improvement would be welcome and we wouldn't have to worry about backwards compatibility.

Reply:
Created: https://github.com/pravega/pravega/issues/4887